### DISCUSSION TITLE
Query parameters from Depends do not show description in docs

### DISCUSSION THREAD

**Discussion by jdratlif** [fastapi/fastapi#8634] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel, Field

app = FastAPI()


class CommonParams(BaseModel):
    a: int = Field(
        1,
        description="Description does not show up in docs",
    )
    b: int = Query(2, description="Using query changes nothing")


@app.get("/")
async def get_root(
    params: CommonParams = Depends(),
    q: int = Query(5, description="This description shows up"),
):
    return {
        "a": params.a,
        "b": params.b,
        "q": q,
    }
```

### Description

Open browser, go to docs page, look at parameters for default route.

Note that description shows up for q parameter, but not for a or b. The schema for a and b in openapi.json does not have a description, but q does.

I expect the description for the fields to show up.

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

3.8.11

**Comment by kheealam** [comment]
You can use a simple class for CommonParams. For example:

```python
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel, Field

app = FastAPI()


class CommonParams:
    def __init__(
        self,
        b: int = Query(2, description="Using query changes nothing"),
        a: int = Query(5, description="This description shows up"),
    ):
        self.b = b
        self.a = a

@app.get("/")
async def get_root(
    params: CommonParams = Depends(),
    q: int = Query(5, description="This description shows up"),
):
    return {
        "a": params.a,
        "b": params.b,
        "q": q,
    }
```

**Comment by jdratlif** [comment]
This is a workaround, and I see that it works with python dataclasses, too, but I still think it should work with pydantic models.

```
from dataclasses import dataclass

from fastapi import Depends, FastAPI, Query

app = FastAPI()


@dataclass
class CommonParams:
    a: int = Query(1, description="describes A")
    b: int = Query(2, description="Describes B")


@app.get("/")
async def get_root(
    params: CommonParams = Depends(),
    q: int = Query(5, description="This description shows up"),
):
    return {
        "a": params.a,
        "b": params.b,
        "q": q,
    }
```

**Comment by dotX12** [comment]
https://github.com/dotX12/pyfa-converter
Solves the problem.

**Comment by long2ice** [comment]
By reading source code I found solution:
```py
class CommonParams(BaseModel):
    a: int = Field(
        Query(...,example=1,description="Description does not show up in docs",
    )
    b: int = Query(
        Query(..., example=2, description="Using query changes nothing")
    ) # this is also ok
```

The code is here: https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L361

Which is not so obvious, but really effective.

**Comment by LanceMoe** [comment]
> By reading source code I found solution:
> 
> ```python
> class CommonParams(BaseModel):
>     a: int = Field(
>         Query(...,example=1,description="Description does not show up in docs",
>     )
>     b: int = Query(
>         Query(..., example=2, description="Using query changes nothing")
>     ) # this is also ok
> ```
> 
> The code is here: https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L361
> 
> Which is not so obvious, but really effective.

A nice workaround, but still has issues in the case of str's enum.

```python
class Region(str, Enum):
    CHINA = 'CN'
    JAPAN = 'JP'

class RegionParams(BaseModel):
    region: Region = Field(
        Query(
            default=Region.CHINA,
            enum=Region,
            description='The region of the zip code.',
        )
    )
```
The description displays fine, but the enum's available values are corrupted.

![middle_img_v2_015e2d8c-9a3e-4542-baf0-63daf4b7500h](https://user-images.githubusercontent.com/18505474/173225699-b97d44e7-da33-4b7e-a822-b49490e5fb3f.png)

![image](https://user-images.githubusercontent.com/18505474/173225724-52a0a3c5-d086-487e-a94e-352f7274433a.png)


In the case of enum, I try to use the following method, which can be bypassed temporarily.

```python
class RegionParams(BaseModel):
    region: Region = Field(
        Query(
            default=Region.CHINA,
            description='The region of the zip code.',
        ),
        enum=Region,
    )
```

Although it can be bypassed in this way, it is still not ideal. I think this is a FastAPI bug, is there a way to fix it?

**Comment by mora123456** [comment]
> By reading source code I found solution:
> 
> ```python
> class CommonParams(BaseModel):
>     a: int = Field(
>         Query(...,example=1,description="Description does not show up in docs",
>     )
>     b: int = Query(
>         Query(..., example=2, description="Using query changes nothing")
>     ) # this is also ok
> ```
> 
> The code is here: https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L361
> 
> Which is not so obvious, but really effective.

I'm not sure this is the perfect solution.
fastapi does have a conversion mechanism from query params to a model, but if you want to use this model in your code and create an instance of this model,
i.e. `params = CommonParams()`:

1.  It will not fail as opposed to:  
```
class CommonParams(BaseModel):
    a: int
    b: int
```
2. The fields in the model will be of type `Query`

I think there should be a better solution for this issue

**Comment by iudeen** [comment]
I think https://github.com/tiangolo/fastapi/pull/4573 is a relevant fix.

**Comment by escopeland** [comment]
@jdratlif  is right: *"...it should work with Pydantic models."*

There are really two issues, though:
1. the Swagger UI should work, as the OP noted; and,
2. an API must sometimes distinguish between two kinds of *optional* query parameter:
    (i) one sent ***explicitly*** in a request to the API with a value equal to its default value, and
    (ii) another ***implicitly*** set to its default value when absent from the request to the API.

Neither the simple `class` solution in the [docs](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/) nor the one suggested by @kheealam on March 22 [above](https://github.com/tiangolo/fastapi/discussions/8634#discussioncomment-5153132) addresses the second issue. Simple class instances can't make the necessary distinction.

But a Pydantic model can, with its `dict()` method's `exclude_unset` and `exclude_defaults` parameters.

So it is perhaps surprising that the solution proposed by @long2ice on June 22 [above](https://github.com/tiangolo/fastapi/discussions/8634#discussioncomment-5153136) doesn't address the issue, either (and it breaks `BaseModel.dict()`'s `exclude_defaults` parameter, too).

The reason is in the `else` clause in `fastapi.dependences.utils.request_params_to_args`,

```
            else:
                values[field.name] = deepcopy(field.default)
```
where `values`, even though not included in the API request, nonetheless gets set explicitly and *then* (not shown here) later passed to the `dependant`, which in this case is the `__init__` method of a Pydantic model. The result is that every field in the Pydantic model is treated as having been explicitly set. 

Can anything be done? Or must we rewrite our query-based `GET` requests instead as `POST` requests (to another URI) in which we insert the query parameters into the request body?

**Comment by alexandercronin** [comment]
This appears to be not a fully supported function yet https://github.com/tiangolo/fastapi/discussions/8348#discussioncomment-9874851

**Comment by YuriiMotov** [comment, marked as answer]
Since FastAPI 0.115.0 you can officially define Query parameters using Pydantic models.
Now use `Query` instead of `Depends`:

```python
from fastapi import FastAPI, Query
from pydantic import BaseModel, Field

app = FastAPI()


class CommonParams(BaseModel):
    a: int = Field(
        1,
        description="Description DOES show up in docs",
    )
    b: int = Query(2, description="Using query also works")


@app.get("/")
async def get_root(
    params: CommonParams = Query(),
    q: int = Query(5, description="This description shows up"),
):
    return {
        "a": params.a,
        "b": params.b,
        "q": q,
    }

```
